### DISCUSSION TITLE
Access to bones and changing their transformation

### DISCUSSION THREAD

**Discussion by fvito** [google/filament#4241] [Q&A]
Hi, [in this sceneform fork](https://github.com/ThomasGorisse/sceneform-android-sdk) I'm trying to integrate the ARCore augmented faces and port the ARCore augmented faces sample, but I'm having some issues getting the bone translation to work. I know filament support bone based animations, but I was not able to find anything about modifying them at runtime without an animation.

At the moment, I'm only able to get the entity ID that gets assigned to the bone. Is there any other way to get the bone information in filament?. I saw that the parser library does support parsing bones, but I was not able to find what the filament asset does with that information.

I was trying to apply transformation of 1 entity to another (via TransformManager) but it did not work, I assume I made an error in terms of transform space. In which space is the entity transformation?

If I modify the transform of a bone entity, will that also affect the mesh that's connected to the bone, or do I need to use something else (like setBonesAsMatrices)?

Thanks, any help is much appreciated

**Comment by romainguy** [comment]
Bones are not represented by entities. `setBonesAsMatrices` and `setBonesAsQuaternions` are what you are looking for.

**Comment by miinso** [comment]
@fvito If you have to use `TransformManager`, you modify the transform node not the bone itself, then call `updateBoneMatrices` yourself.

https://github.com/google/filament/blob/24b88219fa6148b8004f230b377f163e6f184d65/libs/gltfio/src/Animator.cpp#L293

Take [`RiggedSimple.glb`](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/RiggedSimple) for instance, you can modify the transformation data of a node named `Bone.001`.

![rigged_simple_bone](https://user-images.githubusercontent.com/46559594/124379399-b7929c80-dcf1-11eb-84d5-a793573ea29c.png)

Like so:
```kotlin
inner class FrameCallback : Choreographer.FrameCallback {
    private val startTime = System.nanoTime()

    override fun doFrame(frameTimeNanos: Long) {
        choreographer.postFrameCallback(this)

        val elapsedTimeSeconds = (frameTimeNanos - startTime).toDouble() / 1_000_000_000
        val tcm = modelViewer.engine.transformManager

        modelViewer.asset?.getFirstEntityByName("Bone.001")?.let { entity ->
            val boneTransformComponent = tcm.getInstance(entity)
            val transform = rotation(Float3(0.0f,0.0f,1.0f), sin(elapsedTimeSeconds.toFloat()) * 20)
            tcm.setTransform(boneTransformComponent, transpose(transform).toFloatArray())
        }

        modelViewer.animator?.updateBoneMatrices()

        modelViewer.render(frameTimeNanos)
    }
}
```
and you get this:
![riggedsimple_simulated](https://user-images.githubusercontent.com/46559594/124379883-967f7b00-dcf4-11eb-9445-1e973e0b6d56.gif)

This works because `updateBoneMatrices` invokes `setBonesAsMatrices` or `setBonesAsQuaternions` eventually.

https://github.com/google/filament/blob/24b88219fa6148b8004f230b377f163e6f184d65/libs/gltfio/src/Animator.cpp#L320

**Reply by fvito** [reply]
Thank you so much @61315 , this was exactly what I was looking for. 
And also thank you @romainguy for clarifying the part about entities and bones

**Comment by ThomasGorisse** [comment]
Link to our PR : https://github.com/ThomasGorisse/sceneform-android-sdk/pull/93#issuecomment-873618493
